src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -431,28 +432,137 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
+                                        break;
+                                    case "CallConvThiscall":
+                                        actualCallKind = Cci.CallingConvention.ThisCall;
+                                        actualUnmanagedCallingConvetionTypes = ImmutableHashSet<INamedTypeSymbolInternal>.Empty;
+                                        break;
+                                    case "CallConvFastcall":
+                                        actualCallKind = Cci.CallingConvention.FastCall;
+                                        actualUnmanagedCallingConvetionTypes = ImmutableHashSet<INamedTypeSymbolInternal>.Empty;
+                                        break;
+                                    default:
+                                        goto outerDefault;
+                                }
+                                break;
+
+                            default:
+outerDefault:
+                                actualCallKind = Cci.CallingConvention.Unmanaged;
+                                actualUnmanagedCallingConvetionTypes = unmanagedCallingConventionTypes;
+                                break;
+                        }
+                    }
+
+                    // The rules for matching a calling convention are:
+                    // 1. The CallKinds must match exactly
+                    // 2. If the CallKind is Unmanaged, then the set of calling convention types must match exactly, ignoring order
+                    //    and duplicates. We already have both sets in a HashSet, so we can just ensure they're the same length and
+                    //    that everything from one set is in the other set.
+
+                    if (actualCallKind.HasUnknownCallingConventionAttributeBits() || !actualCallKind.IsCallingConvention(expectedConvention.CallKind))
+                    {
+                        results[i] = makeWrongCallingConvention(result);
+                        continue;
+                    }
+
+                    if (expectedConvention.CallKind.IsCallingConvention(Cci.CallingConvention.Unmanaged))
+                    {
+                        switch (expectedConvention.UnmanagedCallingConventionTypes, actualUnmanagedCallingConvetionTypes)
+                        {
+                            case ({ IsEmpty: true }, { IsEmpty: true }):
+                                continue;
+
+                            case ({ IsEmpty: true }, _) or (_, { IsEmpty: true }):
+                            case ({ Count: var expectedCount }, { Count: var actualCount }) when expectedCount != actualCount:
+                                results[i] = makeWrongCallingConvention(result);
+                                continue;
+                        }
+
+                        foreach (var expectedModifier in expectedConvention.UnmanagedCallingConventionTypes)
```

Comment:
>foreach (var expectedModifier in expectedConvention.UnmanagedCallingConventionTypes) [](start = 24, length = 84)

Consider using:
```C#
if (!actualUnmanagedCallingConvetionTypes.SetEquals(expectedConvention.UnmanagedCallingConventionTypes))
{
    results[i] = makeWrongCallingConvention(result);
}
```

Reply:
We can't use `SetEquals`, these collections are of different types.